src/main/java/com/google/sps/utilities/RealTutorSessionDatastore.java

```diff
@@ -92,6 +92,66 @@ public void addTutorSession(String tutorEmail, String studentEmail, TutorSession
         return getScheduledSessions(datastore, "student", email);
     }
 
+    /**
+    * Adds the given rating to a tutor session and updates the tutor's overall rating.
+    * @return boolean, true if session was rated successfully, false otherwise
+    */
+    @Override
+    public boolean rateTutorSession(long sessionId, int rating) {
+
+        boolean success = true;
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        TransactionOptions options = TransactionOptions.Builder.withXG(true);
+        Transaction txn = datastore.beginTransaction(options);
+
+        try {
+
+            Key sessionKey = KeyFactory.createKey("TutorSession", sessionId);
+
+            Entity sessionEntity = datastore.get(txn, sessionKey); 
+
+            sessionEntity.setProperty("rated", true);
+            sessionEntity.setProperty("rating", true);
+
+            String tutorEmail = (String) sessionEntity.getProperty("tutorEmail");
+
+            updateTutorRating(datastore, txn, tutorEmail, rating);
+
+            datastore.put(txn, sessionEntity);
+            
+        } catch (EntityNotFoundException e) {
+            success = false;
+        } finally {
+          if (txn.isActive()) {
+            txn.rollback();
```

Comment:
What happens in the finally block here?

Reply:
This basically just ends the transactions. Even if there was an error somewhere in the try block, the finally statement makes sure the transaction ends.